### PR TITLE
#74 - Add ispolynomial

### LINKED ISSUE #74: Add ispolynomial trait
(Useful to dispatch on algorithms that are special for polynomial ODEs.)

### PR DESCRIPTION
**Pull request by mforets**
Closes #74.

### CODE REVIEW

**Comment by schillic** [comment]
I would say that the semantics of `ispolynomial` should return `true` for a linear system, etc.

**Comment by blegat** [comment]
> I would say that `ispolynomial` should return `true` for a linear system, etc.

Maybe `isalgebraic` then ?

**Comment by mforets** [comment]
yes, `isalgebraic` for [algebraic variety](https://en.wikipedia.org/wiki/Algebraic_variety#Affine_varieties) seems like a better match.

**Comment by mforets** [comment]
This is ready for review.

**Comment by schillic** [comment]
Error [here](https://travis-ci.org/JuliaReach/MathematicalSystems.jl/jobs/502328842#L621):
```
WARNING: both LazySets and MathematicalSystems export "ResetMap"; uses of it in module Main must be qualified
Reset map: Error During Test at /home/travis/build/JuliaReach/MathematicalSystems.jl/test/maps.jl:144
  Got exception outside of a @test
  UndefVarError: ResetMap not defined
  Stacktrace:
   [1] top-level scope at /home/travis/build/JuliaReach/MathematicalSystems.jl/test/maps.jl:145
```

File: src/abstract.jl [hunk 1]
```diff
@@ -87,6 +87,13 @@ For example, polynomial systems can be nonlinear; hence `isaffine` is `false`.
 """
 function isaffine(::AbstractSystem) end
 
+"""
+    isalgebraic(s::AbstractSystem)
+
+Specifies if the dynamics of system `s` is specified by polynomial equations.
```
Comment (schillic) [inline src/abstract.jl]: Please clarify this definition.

**Comment by schillic** [comment]
I do not fully understand the terminology. When does a system have "algebraic dynamics? Why does a `LinearSystem` not have algebraic dynamics, but a `PolynomialSystem` of degree 1 has it?
[Here](https://math.stackexchange.com/a/187566) there is a Wikipedia link, which redirects to [arithmetic dynamics](https://en.wikipedia.org/wiki/Arithmetic_dynamics). Are you sure this is the right term?

**Comment by mforets** [comment]
> I would say that ispolynomial should return true for a linear system, etc.

`ispolynomialtype` then?

**Comment by schillic** [comment]
Actually we already have this semantic inconsistency (`islinear` returns `false` for `AffineContinuousSystem`s even if the additive dynamics is zero). So I give up and you can go with `ispolynomial` or `isalgebraic`.
But I think that this confusion should be prevented with good documentation that these methods do not express system semantics but rather type categories.

**Comment by mforets** [comment]
Yes, see the 2nd paragraph in the [Notes section of that method](https://juliareach.github.io/MathematicalSystems.jl/latest/lib/methods/#MathematicalSystems.islinear-Tuple{AbstractSystem}).

Perhaps `isfootype` is better, i believe, but that amounts to changing the existing methods, so i that would be a separate PR.

**Comment by schillic** [comment]
> Perhaps isfootype is better, i believe, but that amounts to changing the existing methods, so i that would be a separate PR.

I would not change all names. Just stay consistent and everything is fine.

**Pull request merged**